---
title: "R Notebook"
output: html_notebook
---

```{r setup, include=FALSE}
## load in librarys and packages
library("readxl")
library(tidyr)
library(dplyr)
library(arsenal)
library(knitr)
library(VennDiagram)
library(stringr)
library("writexl")
knitr::opts_chunk$set(echo = TRUE)
library(xlsx)
# library(dtutils)
```

## read excel
```{r}
## reused from: https://stackoverflow.com/questions/12945687/read-all-worksheets-in-an-excel-workbook-into-an-r-list-with-data-frames/12945838#12945838
library(readxl)
read_excel_allsheets <- function(filename, tibble = FALSE) {
  # I prefer straight data.frames
  # but if you like tidyverse tibbles (the default with read_excel)
  # then just pass tibble = TRUE
  sheets <- readxl::excel_sheets(filename)
  x <-
    lapply(sheets, function(X)
      readxl::read_excel(filename, sheet = X))
  if (!tibble)
    x <- lapply(x, as.data.frame)
  names(x) <- sheets
  x
}
## reused from: https://stackoverflow.com/questions/17499013/how-do-i-make-a-list-of-data-frames/24352576#24352576

GETDF_FROMLIST <- function(DF_LIST, ITEM_LOC) {
  DF_SELECTED <- DF_LIST[[ITEM_LOC]]
  DF_SELECTED <- tibble::as_tibble(DF_SELECTED)
  return(DF_SELECTED)
}

```

## locations
```{r}
## Give the location of the .xlsx file you want to use
EXCEL_location <- "//home//sanne//bioinf//afstuderen//DATA//RUMC//allmutdiagr1_together_3germlinefiltersapplied_voorsanne.xlsx"
## Give the output location where the output files need to go. 
## Remember the // at the end
OUTPUT_location <- "//home//sanne//bioinf//afstuderen//ANALYSIS//lollipops//"
## Use the mysheets function to read the .xlsx file in as a large list
mysheets <- read_excel_allsheets(EXCEL_location)
```

## Lollipop code
```{r}
lollipop_function <- function(Gene_name, mysheets, OUTPUT_location) {
  Gene_name_allsheets = data.frame()
  for (i in 1:length(mysheets)) {
    # print(i)
    Gene_name_allsheets_row = data.frame()
    sheet <- GETDF_FROMLIST(mysheets, i)
    Gene_name_allsheets_row <-
      subset(sheet , `Gene name` == Gene_name)
    # print(Gene_name_allsheets_row)
    Gene_name_allsheets_row <- Gene_name_allsheets_row %>%
      select(Chromosome, `Start position`, `End position`,
           Reference, Variant, `Gene name`)
    Gene_name_allsheets <-
      rbind(Gene_name_allsheets, Gene_name_allsheets_row)
    # print(Gene_name_allsheets)
  }
  # return(sheet)
  lollipop_data <- Gene_name_allsheets %>%
    select(Chromosome, `Start position`, `End position`,
           Reference, Variant, `Gene name`)
  newheaders <-
    c("Chromosome",
      "Start_Position" ,
      "End_Position",
      "Reference_Allele",
      "Variant_Allele",
      "Hugo_Symbol")
  filenames <-
    paste0(OUTPUT_location, "lollipop_data_dynamisch_", Gene_name, ".txt")
  colnames(lollipop_data) <- newheaders
  write.table(lollipop_data, filenames, na = "NA", sep = "\t", 
              row.names = FALSE, quote = FALSE)
  return(lollipop_data)
}

```

## Get all the hits of this gene and get the right columns for the lollipops
```{r}
## Use the lollipop_function function, give the gene name, the large list with 
## excel data and the output location
lollipop_myc <- lollipop_function("MYC", mysheets, OUTPUT_location)
```
